### ISSUE TITLE
Raft Auto-Snapshotting

### ISSUE THREAD

**Issue by cipherboy** [openbao/openbao#795]
Forking this from #578:

@zarganum said:

> @cipherboy I cannot give commitments but let's first define the feature:)
> Do you have any previous discussions on its scope and use cases?
> Will it be primarily used by single-node deployments (rather than clustered)?
> Should the snapshot be triggered internally (by configured schedule) or externally (e.g. by SIGUSR1)? Signals would limit to POSIX only (not Windows compatible).
> Should the post-snapshot command be run to sync the fresh snapshot to some external storage? If yes, should the bao process track its exit status and abort execution after specified timeout reached?

The API docs for Vault Enterprise are here: https://developer.hashicorp.com/vault/api-docs/system/storage/raftautosnapshots

In short, there seems to be various backends to write the snapshot into (I'd probably start with `storage_type=local` and let others build additional endpoints later). The `bao operator raft snapshot save` command is probably relevant here: https://openbao.org/docs/commands/operator/raft/#snapshot -- calling https://openbao.org/api-docs/system/storage/raft/#take-a-snapshot-of-the-raft-cluster. Except, rather than streaming it to the CLI, we'd obviously save it locally or to one of the other storage backends.

It seems the "command to run" isn't actually used here at all. I'd be more inclined to let another process monitor the snapshot directory than attempting to launch it ourselves...

**Comment by zarganum** [comment]
@cipherboy let me play the devil's advocate role at this point:)

Pls. correct me if I am wrong, but with the current API/CLI and external helpers, it is pretty much possible to deliver the automatic snapshotting to virtually any storage backend possible... E.g. run it via cron or Windows scheduler, does not matter. The snapshot helper can even extract the storage credentials from an unsealed Bao instance (a sealed one is probably unchanged and thus may not require storage sync, right?).

I cannot say the value of autosnapshot **inside** Bao itself is zero. But... what's the motivation, just "enterpriseness"?
How much code will have to be brought into the project for a proper **testing** with all of the backends?

**Comment by zarganum** [comment]
Meanwhile, how are below use cases addressed:
- verify integrity of a snapshot after "long-range" file transfer operations
- protect the running system against malicious or unauthorized snapshots

Is it possible to use existing metadata to verify the particular snapshot is indeed consistent and belongs to the currently running system?

**Comment by siepkes** [comment]
I'll add a reference to this Vault issue which has quite a long discussion in it on backups: https://github.com/hashicorp/vault/issues/5683 

Aside from auto-snapshotting I think in the end for Vault there was never really a definitive answer if the Vault (raft) snapshot for backups was atomic and consistent. In the end Hashicorp left the issue open but closed the discussion with:

> Hi, folks. I'm going to lock commenting down for now while we in the Community team look into this. Thanks for your patience.

That was in 2022 but they don't seem to have taken further action on it.

Is getting an "atomic and consistent" snapshot something which OpenBao might be able to provide thanks to the new transaction API?

Personally I used the following for creating snapshots at regular intervals and uploading them off-site: https://github.com/Argelbargel/vault-raft-snapshot-agent

**Comment by cipherboy** [comment]
@siepkes Yes, my understanding is the existing snapshot API is consistent today, w.r.t. never having a partial log entry and so it will always be restorable. But finishing up #607 will mean any full Raft log snapshot will be consistent.

**Comment by zarganum** [comment]
@siepkes thank you for bringing the VRSA here! Pls. confirm if you do use VRSA with Bao indeed?

@cipherboy see how many external storage options does VRSA have? If I may, that was my "blind personal expectation" and I'm glad VRSA is here. With such a diversity, [plugins](https://openbao.org/docs/plugins) might be the best strategy but unfortunately I don't see how can this fits into one of `secret`, `auth`, or a `database`... unless this is a `database` :)
I also see the motivation of VRSA is original Vault compatibility that may suffer in case of a Bao plugin.

Bottom line: as long as Bao snapshot API is _consistent_, the VRSA-powered Bao appears ahead of its ancestor in the story of automatic snapshots (expect confirmation from @siepkes ).

But may I once again ask: if the Bao snapshot file consistency check available somehow offline or with API?
Use case: ensure the snapshot is valid and can be used indeed, before actual restore operation.

many thanks!
